File: py-engine/validation/schema_validator.py
"""JSON schema validation for Redis messages."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import jsonschema

SCHEMA_DIR = Path(__file__).resolve().parent.parent.parent / "shared" / "schemas"

SchemaName = str  # "market-events" | "execution-orders" | "execution-results"

_schemas: dict[str, dict[str, Any]] = {}


def _load_schema(name: SchemaName) -> dict[str, Any]:
    if name not in _schemas:
        path = SCHEMA_DIR / f"{name}.schema.json"
        _schemas[name] = json.loads(path.read_text())
    return _schemas[name]


class SchemaValidationError(Exception):
    """Raised when a message fails schema validation."""

    def __init__(self, schema_name: str, errors: list[str]) -> None:
        self.schema_name = schema_name
        self.errors = errors
        super().__init__(f"Schema validation failed ({schema_name}): {'; '.join(errors)}")


def validate(schema_name: SchemaName, data: Any) -> tuple[bool, list[str]]:
    """Validate data against a named schema.

    Returns (valid, errors) tuple.
    """
    schema = _load_schema(schema_name)
    validator = jsonschema.Draft202012Validator(schema)
    errors = [
        f"{'/'.join(str(p) for p in e.absolute_path) or '/'}: {e.message}"
        for e in validator.iter_errors(data)
    ]
    return (len(errors) == 0, errors)


def validate_or_raise(schema_name: SchemaName, data: Any) -> None:
    """Validate data and raise SchemaValidationError on failure."""
    valid, errors = validate(schema_name, data)
    if not valid:
        raise SchemaValidationError(schema_name, errors)

File: ts-executor/src/validation/schema-validator.ts
import Ajv, { type ValidateFunction, type ErrorObject } from 'ajv';
import addFormats from 'ajv-formats';
import { readFileSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));

export interface ValidationResult {
  valid: boolean;
  errors: ErrorObject[] | null;
}

export type SchemaName = 'market-events' | 'execution-orders' | 'execution-results';

const SCHEMA_DIR = resolve(__dirname, '../../../shared/schemas');

const ajv = new Ajv({ allErrors: true, strict: false, validateSchema: false });
addFormats(ajv);

const validators = new Map<SchemaName, ValidateFunction>();

/** Load or retrieve a cached JSON schema validator by name. */
function getValidator(name: SchemaName): ValidateFunction {
  let validator = validators.get(name);
  if (!validator) {
    const schemaPath = resolve(SCHEMA_DIR, `${name}.schema.json`);
    const schema = JSON.parse(readFileSync(schemaPath, 'utf-8'));
    // Remove $id to avoid ajv caching conflicts across reloads
    const schemaWithoutId = Object.fromEntries(
      Object.entries(schema).filter(([key]) => key !== '$id'),
    );
    validator = ajv.compile(schemaWithoutId);
    validators.set(name, validator);
  }
  return validator;
}

/** Validate data against a named JSON schema. */
export function validate(schemaName: SchemaName, data: unknown): ValidationResult {
  const validator = getValidator(schemaName);
  const valid = validator(data);
  return {
    valid: valid as boolean,
    errors: valid ? null : (validator.errors ?? null),
  };
}

/** Validate data against a named JSON schema, throwing on failure. */
export function validateOrThrow(schemaName: SchemaName, data: unknown): void {
  const result = validate(schemaName, data);
  if (!result.valid) {
    const messages = (result.errors ?? [])
      .map((e) => `${e.instancePath || '/'}: ${e.message}`)
      .join('; ');
    throw new Error(`Schema validation failed (${schemaName}): ${messages}`);
  }
}
